external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp

```diff
@@ -385,53 +419,59 @@ void outlineConvPartOps(Operation *convOp, ArrayRef<Operation *> secondOps,
   }
 
   // Erase the ops we outlined, which should be safe now.
-  for (auto &op : llvm::make_early_inc_range(llvm::reverse(secondOps))) {
+  for (auto &op : llvm::make_early_inc_range(llvm::reverse(trailingOps))) {
     if (op->use_empty()) {
       op->erase();
     }
   }
-  assert(convOp->use_empty() && "expected 'op' to have no uses");
-  convOp->erase();
-  for (auto &op : llvm::make_early_inc_range(frontOps)) {
+  assert(anchorOp->use_empty() && "expected 'op' to have no uses");
+  anchorOp->erase();
+  for (auto &op : llvm::make_early_inc_range(leadingOps)) {
     if (op->use_empty()) {
       op->erase();
     }
   }
 }
 
+} // namespace
+
+namespace {
+
 // Inspired by / adapted from TestSCFIfUtilsPass in
 // test/lib/Transforms/TestSCFUtils.cpp.
 class TosaPartitionPass : public TosaPartitionBase<TosaPartitionPass> {
-public:
-  static bool isZeroAttribute(Attribute value) {
-    if (auto intValue = value.dyn_cast<IntegerAttr>())
-      return intValue.getValue().isNullValue();
-    if (auto fpValue = value.dyn_cast<FloatAttr>())
-      return fpValue.getValue().isZero();
-    if (auto splatValue = value.dyn_cast<SplatElementsAttr>())
-      return isZeroAttribute(splatValue.getSplatValue<Attribute>());
-    if (auto elementsValue = value.dyn_cast<ElementsAttr>())
-      return llvm::all_of(elementsValue.getValues<Attribute>(),
-                          isZeroAttribute);
-    if (auto arrayValue = value.dyn_cast<ArrayAttr>())
-      return llvm::all_of(arrayValue.getValue(), isZeroAttribute);
-    return false;
+  // Special case:  TransposeOp's second operand must be a
+  // constant, which means we must include it too if we include
+  // the TransposeOp.  "ops" here may be either leadingOps or trailingOps.
+  void specialCaseForTranspose(Operation *op, SetVector<Operation *> &ops) {
+    auto *operand = op->getOpOperand(1).get().getDefiningOp();
+    ops.insert(operand);
   }
 
-  static bool isConstantZero(Operation *op) {
-    // test for zero
-    if (auto cst = dyn_cast<arith::ConstantOp>(op)) {
-      return isZeroAttribute(cst.getValue());
-    }
-    return false;
+public:
+  TosaPartitionPass() = default;
+
+  virtual bool isAnchorOp(Operation *op) {
+    return isa<tosa::Conv2DOp,tosa::MatMulOp,tosa::DepthwiseConv2DOp>(op);
+  }
+  virtual bool isLeadingOp(Operation *op) {
+    return isConstantZero(op) || isFuseableOp(op);
+  }
+  virtual bool isTrailingOp(Operation *op) {
+    return isFuseableOp(op);
+  }
+  virtual std::string attributeName() {
+    return "kernel";
   }
 
-  void traceInputs(Operation *op, SmallVector<Operation *> &predecessors,
+  void traceInputs(Operation *op, SetVector<Operation *> &predecessors,
```

Comment:
Leading ops (predecessors) became SetVector<> so I could abstract the TransposeOp special case.  Doesn't matter much besides that.

Reply:
Yes, it preserves the order of addition.  That's one reason I used it originally.